src/source.cpp

```diff
@@ -430,15 +517,30 @@ MeshSource::MeshSource(pugi::xml_node node)
 
 SourceSite MeshSource::sample(uint64_t* seed) const
 {
-  // sample location and element from mesh
-  auto mesh_location = space_->sample_mesh(seed);
-
-  // Sample source for the chosen element
-  int32_t element = mesh_location.first;
-  SourceSite site = source(element)->sample(seed);
+  // Sample the CDF defined in initialization above
+  int32_t element = space_->sample_element_index(seed);
+
+  // Sample position and apply rejection on spatial domains
+  Position r;
+  while (true) {
+    r = space_->mesh()->sample_element(element, seed);
+    if (this->satisfies_spatial_constraints(r)) {
```

Comment:
OK, I see why you'd want to separate this out as its own function now. It's nice to pre-screen out some rejection sites before passing back to the `Source` class's rejection loop.